### PR TITLE
DD resort refactor

### PR DESCRIPTION
**Pull request by fweik**
Description of changes:
 - DD resort refactored into multiple functions and simplified.
 - Fixed layered with full periodic boundaries (hopefully)
 - Extended pair finding test
 - Resort now works with folded coords
 - Moved local sort to cells level
 - Moved particle position reset from short-range loop (?) to sorting code.
 - Minor performance improvements.

### CODE REVIEW

File: src/core/cells.cpp [hunk 1]
```diff
@@ -302,6 +302,73 @@ int cells_get_n_particles() {
...
 
+namespace {
+void fold_and_reset(Particle &p) {
+  fold_position(p.r.p, p.l.i);
+
+  for (int i = 0; i < 3; i++) {
+    p.l.p_old[i] = p.r.p[i];
+  }
+}
+
```
Comment (KaiSzuttor) [inline src/core/cells.cpp:309]: could you please add docstrings to the new functions?

File: src/core/cells.cpp [hunk 2]
```diff
@@ -311,24 +378,40 @@ void cells_resort_particles(int global_flag) {
...
 
 #ifdef ADDITIONAL_CHECKS
   /* at the end of the day, everything should be consistent again */
+  /* These have to be run before we put the particles that we already
+   * know are displaced back, otherwise they will trigger different errors.*/
   check_particle_consistency();
   check_particle_sorting();
 #endif
 
+  if (0 != displaced_parts.n) {
```
Comment (KaiSzuttor) [inline src/core/cells.cpp:406]: this could be a function

File: src/core/domain_decomposition.cpp [hunk 3]
```diff
@@ -904,243 +841,117 @@ void dd_topology_release() {
...
+    } else {
+
+      append_unindexed_particle(&rest, std::move(src.part[i]));
+    }
+  }
+
+  realloc_particlelist(&src, src.n = 0);
+}
+
+void move_left_or_right(ParticleList &src, ParticleList &left,
```
Comment (KaiSzuttor) [inline src/core/domain_decomposition.cpp:775]: please add a docstring here too

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

**Comment by fweik** [comment]
@KaiSzuttor

**Comment by fweik** [comment]
@RudolfWeeber, we maybe should also look at some benchmarks to check that I did not screw up too much. Also I think we should find some permanent solution for the NPT test, it has a false positive every time the order the PRNG is applied changes. This is not what it is supposed to check.

**Comment by fweik** [comment]
@RudolfWeeber ready from my side. The NPT issue remains, but I think it's a false positive.

**Comment by RudolfWeeber** [comment]
I think, this PR can be merged.

**Pull request merged**